flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java

```diff
@@ -869,41 +859,40 @@ public ApplicationReport startAppMaster(
 		FsPermission permission = new FsPermission(FsAction.ALL, FsAction.NONE, FsAction.NONE);
 		fs.setPermission(yarnFilesDir, permission); // set permission for path.
 
+		Path remoteYarnSiteXmlPath = null;
+		File f = new File(System.getenv("YARN_CONF_DIR"), Utils.YARN_SITE_FILE_NAME);
+		LOG.info("Adding Yarn configuration {} to the AM container local resource bucket", f.getAbsolutePath());
+		Path yarnSitePath = new Path(f.getAbsolutePath());
+		remoteYarnSiteXmlPath = setupSingleLocalResource(
+			Utils.YARN_SITE_FILE_NAME,
+			fs,
+			appId,
+			yarnSitePath,
+			localResources,
+			homeDir,
+			"");
```

Comment:
This is not what I proposed. We should not add the `yarn-site.xml` per default to all Yarn deployments. Instead we should only add it if we are using the `YarnTestBase`. Therefore, I suggest to add 
```
yarnClusterDescriptor.addShipFiles(Arrays.asList(flinkLibFolder, yarnSiteXMLFile));
```

to `YarnTestBase#createYarnClusterDescriptor`.

Reply:
@tillrohrmann Unfortunately, not all yarn related test (e.g. YARNSessionFIFOITCase#testDetachedMode) use `YarnTestBase#createYarnClusterDescriptor`, so I can not only add yarn-site.xml explicitly in YarnTestBase

Reply:
I guess one could then refactor the `YarnTestBase#Runner` to instantiate the `FlinkYarnSessionCli` with a `Supplier<YarnClusterDescriptor>` which could setup the `YarnClusterDescriptor` accordingly.

This might be out of scope for this PR but maybe this test case should not test command line parsing and deploying a Yarn application. Instead one could this split up into two tests: One tests the command line parsing (`CliFrontend`) and the other test tests the detached mode.

Reply:
> I guess one could then refactor the YarnTestBase#Runner to instantiate the FlinkYarnSessionCli with a Supplier<YarnClusterDescriptor> or factory which could setup the YarnClusterDescriptor accordingly.

@tillrohrmann IMHO, this is not a good solution as well. Because it would also introduce different code path for test code and production code in `FlinkYarnSessionCli`. 
I totally agree that we need to refactor the flink client api, currently the command line parsing and cluster deployment is coupled together closely, which make it difficult to do unit testing (The test code in flink-yarn-tests module is not readable to me, which imply the issues of current flink api design).  I do have a plan to refactor the flink client api, but it would involve a large amount of work. 
For this PR, it is just to fix the YarnMiniCluster integration test bug so that downstream project can do integration test with flink. I think we can just fix the bug without touching the client api (keep `IN_TESTS`) for now. The refactoring of flink client api could be done in other PRs. Does it make sense to you ?

Reply:
Ok, I agree that this would widen the scope quite a bit. So the idea would then simply to always ship the yarn-site.xml file if `IN_TESTS` is set, right? Additionally, we will write the `yarn-site.xml` to the `test-classes` directory.

Please open the corresponding JIRA issues to properly fix this in the future.

Reply:
That's correct, I have created https://issues.apache.org/jira/browse/FLINK-12422 for the followup work.

Reply:
Thanks!